lua/orgmode/org/hyperlinks.lua

```diff
@@ -2,33 +2,79 @@ local Files = require('orgmode.parser.files')
 local utils = require('orgmode.utils')
 local Hyperlinks = {}
 
-function Hyperlinks.find_by_custom_id_property(base, skip_mapping)
-  local headlines = Files.get_current_file():find_headlines_with_property_matching('CUSTOM_ID', base:sub(2))
-  if skip_mapping then
+local function get_file_from_context(ctx)
+  return (
+      ctx.hyperlinks and ctx.hyperlinks.filepath and Files.get(ctx.hyperlinks.filepath, true)
+      or Files.get_current_file()
+    )
+end
+
+local function update_hyperlink_ctx(ctx)
+  if not ctx.line then
+    return
+  end
+
+  -- TODO: Support text search, see here [https://orgmode.org/manual/External-Links.html]
+  local hyperlinks_ctx = {
+    filepath = false,
+    headline = false,
+    custom_id = false,
+  }
+
+  local file_match = ctx.line:match('file:(.-)::')
+  if file_match and Files.get(file_match) then
+    hyperlinks_ctx.filepath = Files.get(file_match).filename
+    hyperlinks_ctx.headline = ctx.line:match('file:.-::(%*.-)$')
+
+    if not hyperlinks_ctx.headline then
+      hyperlinks_ctx.custom_id = ctx.line:match('file:.-::(#.-)$')
+    end
+
+    ctx.base = hyperlinks_ctx.headline or hyperlinks_ctx.custom_id or ctx.base
+  end
+
+  ctx.hyperlinks = hyperlinks_ctx
+end
+
+function Hyperlinks.find_by_filepath(_)
+  local filenames = Files.filenames()
```

Comment:
Add additional filtering to this:
```lua
function Hyperlinks.find_by_filepath(ctx)
  local filenames = Files.filenames()
  local file_base = ctx.base:gsub('^file:', '')
  if vim.trim(file_base) ~= '' then
  filenames = vim.tbl_filter(function(f)
    return f:find('^'..file_base)
  end, filenames)
  end

  -- Outer checks already filter cases where `ctx.skip_add_prefix` is truthy,
  -- so no need to check it here
  return vim.tbl_map(function(path)
    return 'file:' .. path
  end, filenames)
end
```
Idea is to grab everything after the base `file:` and filter only filenames that are matching the input, like I explained in the comment for regex change.